File: part 4/practice/Interpreter.py
from Lexer import Lexer
from Token import Token


class Interpreter(object):
    def __init__(self, text):
        self.lexer = Lexer(text)
        self.current_token = self.lexer.get_next_token()

    def error(self):
        raise Exception('Error parsing input')

    def compare(self, token_types):
        if self.current_token.type in token_types:
            self.current_token = self.lexer.get_next_token()
        else:
            self.error()

    def term(self):
        """<factor> ::= <whitespace>* <number> <whitespace>*"""

        result = self.current_token
        self.compare([Token.INTEGER])

        return result.value

    def operator(self):
        """<op> ::= "*" | "/" | "+" | "-" """

        op = self.current_token
        self.compare([Token.MULTIPLY, Token.DIVIDE, Token.PLUS, Token.MINUS])

        return op

    def eval(self, left, op, right):
        """<eval> ::= <term> <op> <term>"""

        if op.type == Token.MULTIPLY:
            return left * right
        elif op.type == Token.DIVIDE:
            return left / right
        elif op.type == Token.PLUS:
            return left + right
        elif op.type == Token.MINUS:
            return left - right

    def expr(self):
        """<expression> ::= <term> { <op> <term> }*"""

        result = self.term()

        while self.current_token.type in [Token.MULTIPLY, Token.DIVIDE, Token.PLUS, Token.MINUS]:
            op = self.operator()
            right = self.term()
            result = self.eval(result, op, right)

        return int(result)

File: part 4/practice/Interpreter.cs
using System;
using test.@base;

namespace test.part_4.practice
{
    public class Interpreter : IInterpreter
    {
        Lexer lexer { get; init; }

        public Interpreter(string text)
        {
            lexer = new Lexer(text);
        }

        private static void Error(string text)
        {
            throw new Exception($"Invalid element exception. {text} was not expected.");
        }

        private int Op(int left, string op, int right)
        {
            switch (Char.Parse(op))
            {
                case '+':
                    return left + right;

                case '-':
                    return left - right;

                default:
                    Error(op);
                    return -1;
            }
        }

        private int Term(int idx)
        {
            return int.Parse(lexer.Tokens[idx].Value);
        }

        private static int Eval(int left, string op, int right)
        {
            switch(op)
            {
                case "+":
                    return left + right;

                case "-":
                    return left - right;

                case "*":
                    return left * right;

                case "/":
                    return left / right;

                default:
                    Error(op);
                    return -1;
            }
        }

        public int Eval()
        {
            lexer.Tokenize();

            int result = Term(0);

            for (int i = 1; i < lexer.Tokens.Count();)
            {
                string op = lexer.Tokens[i++].Value;
                int right = Term(i++);
                result = Eval(result, op, right);
            }

            return result;
        }
    }
}
